graphrag/storage/cosmosdb_pipeline_storage.py

```diff
@@ -0,0 +1,300 @@
+# Copyright (c) 2024 Microsoft Corporation.
+# Licensed under the MIT License
+
+"""Azure CosmosDB Storage implementation of PipelineStorage."""
+
+import json
+import logging
+import re
+from collections.abc import Iterator
+from io import BytesIO, StringIO
+from typing import Any
+
+import pandas as pd
+from azure.cosmos import CosmosClient
+from azure.cosmos.partition_key import PartitionKey
+from azure.identity import DefaultAzureCredential
+from datashaper import Progress
+
+from graphrag.logging.base import ProgressReporter
+from graphrag.storage.pipeline_storage import PipelineStorage
+
+log = logging.getLogger(__name__)
+
+
+class CosmosDBPipelineStorage(PipelineStorage):
+    """The CosmosDB-Storage Implementation."""
+
+    _cosmosdb_account_url: str | None
+    _connection_string: str | None
+    _database_name: str
+    _current_container: str | None
+    _encoding: str
+
+    def __init__(
+        self,
+        cosmosdb_account_url: str | None,
+        connection_string: str | None,
+        database_name: str,
+        encoding: str = "utf-8",
+        current_container: str | None = None,
+    ):
+        """Initialize the CosmosDB Storage."""
+        if connection_string:
+            self._cosmos_client = CosmosClient.from_connection_string(connection_string)
+        else:
+            if cosmosdb_account_url is None:
+                msg = (
+                    "Either connection_string or cosmosdb_account_url must be provided."
+                )
+                raise ValueError(msg)
+
+            self._cosmos_client = CosmosClient(
+                url=cosmosdb_account_url,
+                credential=DefaultAzureCredential(),
+            )
+
+        self._encoding = encoding
+        self._database_name = database_name
+        self._connection_string = connection_string
+        self._cosmosdb_account_url = cosmosdb_account_url
+        self._current_container = current_container
+        self._cosmosdb_account_name = (
+            cosmosdb_account_url.split("//")[1].split(".")[0]
+            if cosmosdb_account_url
+            else None
+        )
+        self._database_client = self._cosmos_client.get_database_client(
+            self._database_name
+        )
+
+        log.info(
+            "creating cosmosdb storage with account: %s and database: %s",
+            self._cosmosdb_account_name,
+            self._database_name,
+        )
+        self._create_database()
+        if self._current_container:
+            self._create_container()
+
+    def _create_database(self) -> None:
+        """Create the database if it doesn't exist."""
+        self._cosmos_client.create_database_if_not_exists(id=self._database_name)
+
+    def _delete_database(self) -> None:
```

Comment:
This method is never actually used and is not required by the abstract class. Should we just delete it?